### DISCUSSION TITLE
Support for OpenAPI parameter serialization

### DISCUSSION THREAD

**Discussion by adriangb** [Neoteroi/BlackSheep#209] [Q&A]
Does BlackSheep support for [OpenAPI parameter serialization](https://swagger.io/docs/specification/serialization/) (aside from the defaults)?

**Comment by RobertoPrevato** [comment]
Hi @adriangb
No, currently only defaults are supported. And thank you for bringing this up, this would be an interesting feature. It is possible to define custom binders, so it's open to such extensions, but the interesting point would be to include also the generation of OpenAPI Documentation.

**Reply by adriangb** [reply]
Is there any relationship between binders and the OpenAPI docs? As in, do the binders present an interface that can be used to customize the OpenAPI docs they produce?

I am also curious about how you would handle the typing aspect of this. AFAIK `FromQuery[..., style="deepObject"]` is not legal. You'd have to do something like `Annotated[..., FromQuery(style="deepObject")]` I think.

**Reply by RobertoPrevato** [reply]
@adriangb yes, binders are inspected in the logic that creates OpenAPI Documentation. However, binders themselves have the single responsibility of binding values and are not concerned with documentation - I like this to stay as is.

By the way, regarding the other issue you reported, I just noticed that `Form` binders currently don't work properly because of a silly bug I made in code, see:
https://github.com/Neoteroi/BlackSheep/blob/88679b653334072127830f442300ed5823764fcf/blacksheep/server/bindings.py#L508

```python
class FormBinder(BodyBinder):
    """
    Extracts a model from form content, either
    application/x-www-form-urlencoded, or multipart/form-data.
    """

    handle = FromForm

    @property
    def content_type(self) -> str:
        return "application/json"    # <--- this is a silly bug in the current code
```

I will fix this bug, and support differentiating between `application/x-www-form-urlencoded` and `multipart/form-data` for the generation of documentation.

**Reply by RobertoPrevato** [reply]
To preserve consistency, probably two dedicated binders should be added, one for each type of form mime. The current "form-binder" could generate two possible mimes in the OAD.

**Reply by adriangb** [reply]
awesome. I agree that having two different binders makes sense, and I like your strategy for backwards compatibility.

With respect to using binders for documentation, I agree with you that it is better to keep things to having a single responsibility. However, I have a couple counterpoints. These are questions I am legitimately unsure of, and I'm curious what you think, if you have the time to answer 😃

1. The options mentioned are not just for documentation. They determine how the values are extracted from the request (I guess that is binding in BlackSheep terminology)
2. They are already influencing the OpenAPI documentation
3. What is the alternative? Using seperate markers (in type annotations like BlackSheep or default values like FastAPI)?

**Reply by RobertoPrevato** [reply]
In cases like this one, I think respecting the single responsibility principle is a requisite. Imagine if I made the Binders (currently they only care about extracting a value from a web request) care also about their representation in the current version of OpenAPI Documentation: v3. What if somebody wants to support v2? What when there will be OpenAPI Documentation v4? Will everything happen in the same classes?

There are ways to keep things decoupled so that supporting a older version of OAD or a newer won't require touching the classes that are already doing something important.

Just my opinion, after all there are thousands of different opinions when it comes to programming, so I like to have some myself. 🙂

**Reply by adriangb** [reply]
I generally agree with your opinion, does that mean it is not yours anymore 😛?

Just to clarify though, would you say that the binders currently adhere to the single responsibility principle? It sounds like they are used to generate the docs, even if they themselves are not aware of the OpenAPI at all.

For my toy project, I was thinking of doing something like:

```python
def endpoint(value: Annotated[int, ExtractFromQuery(style="deepObject"), OpenAPIv3QueryParam(style="deepObject")]):    ...
```

Only problem is that this is super verbose! As far as I can tell, there is no way to "compose" these things into some sane default (like `def endpoint(value: FromQuery(style="deepObject"))`, which would be ideal) due to limitations of (1) annotations not supporting `()` syntax directly and (2) https://github.com/python/typing/issues/779

**Reply by RobertoPrevato** [reply]
Yes, in my mind binders today have a single responsibility, and I think this is proven by the fact that the base abstract class `Binder` defines a single abstract method `get_value` that sub-classes must provide.

When it comes to your example, I would recommend to not try to make a generic API for rare scenarios. What I mean is: most of times query strings are used to handle scalar values or arrays (much more often scalar values than arrays, by the way -like "response_mode=form_post"). If the user of the framework has specific needs, enable defining a custom class to bind incoming values.

In my case, this is supported in blacksheep defining a [custom binder](https://www.neoteroi.dev/blacksheep/binders/#defining-a-custom-binder) - which is by the way the same approach found in [ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/advanced/custom-model-binding?view=aspnetcore-6.0).

**Reply by RobertoPrevato** [reply]
By the way, I was planning to add more extensibility points to the current logic that generates OpenAPI Documentation v3, to make it easier to modify the documentation generated for specific parameters / endpoints. I didn't yet simply because time is limited.
